DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/matmul/README.md

```diff
@@ -0,0 +1,323 @@
+
+### Performance
+
+Performance results are based on testing as of March 6, 2023.
+
+> **Note**: Refer to the [Performance Disclaimers](/DirectProgramming/C++SYCL_FPGA/README.md#performance-disclaimers) section for important performance information.
+
+| Device                                            | Throughput
+|:---                                               |:---
+| Intel® PAC with Intel® Arria® 10 GX FPGA          | 75k matrices/s for matrices of size 64 * 64
+| Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) | 90k matrices/s for matrices of size 64 * 64
+
+
+## Key Implementation Details
+
+The matrix multiplication algorithm multiplies two single-precision floating-point matrices *A* (*m × n*) and *B* (*n × p*) to produce matrix *C* (*m × p*). The algorithm is implemented using a systolic array approach, employing an array of *m<sub>PE</sub> × p<sub>PE</sub>* processing elements (PEs), each of which is responsible for computing a dot product corresponding to an element of the output matrix. As a result, this implementation computes the matrix multiplication in tiles of *m<sub>PE</sub> × p<sub>PE</sub>* at a time.
+
+<p align="center">
+  <img src=assets/overview.png />
+</p>
+
+Each iteration, a PE consumes inputs from previous PEs in the same row or column and uses these data to compute a multiply-accumulate. It then passes the inputs along to the next PE. We orchestrate the flow of data through the systolic array such that each PE is responsible for computing one element of the output matrix. Using this approach, it takes a total of *n* iterations to compute one matrix product, where every iteration, we compute a partial sum for each of the elements of the output matrix.
+
+<p align="center">
+  <img src=assets/systolic_array.png />
+</p>
+
+Here, a single PE consists of logic to compute a floating-point multiply-accumulate and registers for storage of intermediate results. 
+
+<p align="center">
+  <img src=assets/PE.png />
+</p>
+
+In our FPGA implementation, the multiply-accumulate operations are computed in a nested loop over the PEs. The loop is fully unrolled to replicate the hardware for each PE. As a result, *m<sub>PE</sub> × p<sub>PE</sub>* multiply-accumulate operations are performed in parallel on the FPGA. The multiply-accumulate arithmetic can be optimally implemented using the FPGA's specialized floating point DSP (Digital Signal Processing) hardware. When set to FP accum mode, we can perform this operation with an II of 1. With this optimization, our FPGA implementation requires *m<sub>PE</sub> × p<sub>PE</sub>* DSPs to compute all the floating-point multiply-accumulate operations. Thus, the matrix tile size is constrained by the total FPGA DSP resources available.
+
+To optimize the performance-critical loop in its algorithm, the design leverages concepts discussed in the following FPGA tutorials:
+
+- **Explicit Pipelining with `fpga_reg`** (fpga_register)
+- **Loop `ivdep` Attribute** (loop_ivdep)
```

Comment:
I'll double check this again. If I recall correctly there was a case where the compiler would report a memory dependency when there wasn't one (since we only write the matrices out to DDR once).